elastica/contact_forces.py

```diff
@@ -725,3 +823,128 @@ def apply_contact(
             self.k,
             self.nu,
         )
+
+
+class RodMeshSurfaceContact(NoContact):
+    """
+    This class is for applying contact forces between rod-mesh_surface.
+    First system is always rod and second system is always mesh_surface.
+
+    Examples
+    --------
+    How to define contact between rod and mesh_surface.
+
+    >>> simulator.detect_contact_between(rod, mesh_surface).using(
+    ...    RodMeshSurfaceContact,
+    ...    k=1e4,
+    ...    nu=10,
+    ...    surface_tol=1e-2,
+    ... )
+
+
+    .. [1] Preclik T., Popa Constantin., Rude U., Regularizing a Time-Stepping Method for Rigid Multibody Dynamics, Multibody Dynamics 2011, ECCOMAS. URL: https://www10.cs.fau.de/publications/papers/2011/Preclik_Multibody_Ext_Abstr.pdf
+    """
+
+    def __init__(self, k: float, nu: float, surface_tol=1e-4):
+        """
+
+        Parameters
+        ----------
+        k: float
+            Stiffness coefficient between the plane and the rod-like object.
+        nu: float
+            Dissipation coefficient between the plane and the rod-like object.
+        surface_tol: float
+            Penetration tolerance between the surface and the rod-like object.
+
+        """
+        super(RodMeshSurfaceContact, self).__init__()
+        # n_faces = faces.shape[-1]
+        self.k = k
+        self.nu = nu
+        self.surface_tol = surface_tol
+
+    def _check_systems_validity(
+        self,
+        system_one: SystemType,
+        system_two: AllowedContactType,
+    ) -> None:
+        """
+        This checks the contact order and type of a SystemType object and an AllowedContactType object.
+        For the RodMeshSurfaceContact class first_system should be a rod and second_system should be a mesh_surface.
+
+        Parameters
+        ----------
+        system_one
+            SystemType
+        system_two
+            AllowedContactType
+        """
+        if not issubclass(system_one.__class__, RodBase) or not issubclass(
+            system_two.__class__, MeshSurface
+        ):
+            raise TypeError(
+                "Systems provided to the contact class have incorrect order/type. \n"
+                " First system is {0} and second system is {1}. \n"
+                " First system should be a rod, second should be a mesh surface".format(
+                    system_one.__class__, system_two.__class__
+                )
+            )
+
```

Comment:
Can you add validity checks for the grid, the grid is created before being passed to the class and depends on both the mesh_surface and the rod. You could use lines 310-322 in `continuum_snake.py` in the [ContinuumSnakeTerrainCase](https://github.com/Ali-7800/PyElastica/tree/dev_snake_contact/examples/ContinuumSnakeTerrainCase) for reference. Essentially you need to check that the given grid is appropriate for the given rod-mesh_surface combination.

Reply:
Right, I will add them.
How and where do you suggest we add the surface_grid, surface_grid_numba functions for the user to create the grid?
One suggestion is contact_utils.py.

Reply:
I think that's fine

Reply:
Okay, I will uncomment it in the next commit.